onnxruntime/core/optimizer/matmul_transpose_fusion.cc

```diff
@@ -0,0 +1,133 @@
+  ORT_ENFORCE(!node_consumers.empty());
+  auto it = count_map.find(target);
+  if (it == count_map.end()) {
+    count_map.insert({target, node_consumers.size() - 1});
+    return node_consumers.size() - 1;
+  } else {
+    count_map[target] -= 1;
+    return count_map[target];
+  }
+}
+
+Status MatmulTransposeFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::deque<onnxruntime::NodeIndex> removed_nodes;
+
+  std::unordered_map<NodeArg*, size_t> consumer_count;
+
+  for (auto node_index : node_topology_list) {
+    auto& node = *graph.GetNode(node_index);
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if ((!graph_utils::IsSupportedOptypeVersionAndDomain(node, "MatMul", {9}) &&
+         !graph_utils::IsSupportedOptypeVersionAndDomain(node, "TransposeMatMul", {9})) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders())) {
+      continue;
+    }
+
+    NodeArg* left_input = node.MutableInputDefs()[0];
+    auto left = IsInputTranspose(graph, *left_input);
+
+    NodeArg* right_input = node.MutableInputDefs()[1];
+    auto right = IsInputTranspose(graph, *right_input);
+
+    if (!left.first && !right.first) {
+      continue;
+    }
+
+    if (left.first) {
+      size_t left_consumers = UpdateConsumerCount(graph, left_input, consumer_count);
+      if (left_consumers == 0)
+        removed_nodes.push_front(left.second->Index());
+      left_input = left.second->MutableInputDefs()[0];
+    }
+
+    if (right.first) {
+      size_t right_consumers = UpdateConsumerCount(graph, right_input, consumer_count);
+      if (right_consumers == 0)
+        removed_nodes.push_front(right.second->Index());
+      right_input = right.second->MutableInputDefs()[0];
+    }
+
+    const std::vector<NodeArg*> input_defs{left_input, right_input};
+    const std::vector<NodeArg*> output_defs{node.MutableOutputDefs()[0]};
+
+    Node& matmul_node = graph.AddNode(graph.GenerateNodeName("MatMul_With_Transpose"),
+                                      "TransposeMatMul",
+                                      "fused MatMul and Transpose ",
+                                      input_defs,
+                                      output_defs, {}, kMSDomain);
+    bool transpose_left = left.first;
+    if (node.OpType() == "TransposeMatMul") {
+      transpose_left ^= static_cast<bool>(node.GetAttributes().at("transA").i());
+    }
+    bool transpose_right = right.first;
+    if (node.OpType() == "TransposeMatMul") {
+      transpose_right ^= static_cast<bool>(node.GetAttributes().at("transB").i());
+    }
+    matmul_node.AddAttribute("transA", (int64_t)transpose_left);
+    matmul_node.AddAttribute("transB", (int64_t)transpose_right);
+    // Assign provider to this new node. Provider should be same as the provider for old node.
```

Comment:
What tests this? The model in the unit test only has the simplest case - one Transpose into a MatMul. Is there any testing of both inputs to the MatMul being Transpose? 

Assuming there's no testing of the code where 'node' is TransposeMatMul due to the issue with matching that op's version. When would that occur in a graph?

Reply:
added a few more UTs

---
In reply to: [397606207](https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397606207) [](ancestors = 397606207)